File: Low Level Design (LLD)/DesignPatterns/2. Behavariol/Composite/Cpp/src/Directory.cpp
#include "../include/Directory.h"
#include <iostream>

Directory::Directory(const std::string& name) : name(name) {}

void Directory::add(FileComponent* component) {
    children.push_back(component);
}

void Directory::showDetails(const std::string& indent) {
    std::cout << indent << "Directory: " << name << std::endl;
    for (auto child : children) {
        child->showDetails(indent + "  ");
    }
}

Directory::~Directory() {
    for (auto child : children) {
        delete child;
    }
}

File: Low Level Design (LLD)/DesignPatterns/2. Behavariol/Composite/Java/Directory.java
import java.util.ArrayList;
import java.util.List;

public class Directory implements FileComponent {
    private String name;
    private List<FileComponent> children = new ArrayList<>();

    public Directory(String name) {
        this.name = name;
    }

    public void add(FileComponent component) {
        children.add(component);
    }

    public void remove(FileComponent component) {
        children.remove(component);
    }

    @Override
    public void showDetails(String indent) {
        System.out.println(indent + "Directory: " + name);
        for (FileComponent child : children) {
            child.showDetails(indent + "  ");
        }
    }
}
